lib/ecto/pools/ownership.ex

```diff
@@ -0,0 +1,206 @@
+defmodule Ecto.Pools.Ownership do
+
+  defmodule Strategy do
+    use Behaviour
+    alias Ecto.Pool
+
+    defcallback ownership_checkout(module, pid) :: Pool.mode
+    defcallback ownership_checkin(module, pid) :: Pool.mode
+  end
+
+  defmodule Supervisor do
+    use Elixir.Supervisor
+
+    def start_link(connection, opts) do
+      Elixir.Supervisor.start_link(__MODULE__, {connection, opts})
+    end
+
+    def init({connection, opts}) do
+      {pool, opts} = Keyword.pop(opts, :ownership_pool)
+      name         = Keyword.fetch!(opts, :pool_name)
+      pool_name    = Module.concat(name, Inner)
+      pool_opts    = Keyword.put(opts, :pool_name, pool_name)
+
+      children = [
+        worker(Ecto.Pools.Ownership.Server, [name, pool, pool_name]),
+        supervisor(pool, [connection, pool_opts])
+      ]
+      supervise(children, strategy: :rest_for_one)
+    end
+  end
+
+  defmodule Server do
+    use GenServer
+
+    @behaviour Ecto.Pool
+    @timeout 5_000
+
+    def start_link(connection, opts) do
+      Supervisor.start_link(connection, opts)
+    end
+
+    def start_link(name, pool, pool_name) do
+      GenServer.start_link(__MODULE__, {pool, pool_name}, [name: name])
+    end
+
+    def ownership_checkout(pool, strategy, timeout \\ @timeout) do
```

Comment:
We can change to return :ok. In the future this function will return a ref to handle handoffs anyway.